File: examples/www/public/assets/scripts/lua/texture.lua
local Texture = {}
Texture.__index = Texture
local ResourceAsync = require("resource_async")

local function load_sync(path)
  local handle = load_texture(path)
  if handle == nil or handle == 0 then
    return nil
  end

  return setmetatable({
    handle = handle,
    x = 0.0,
    y = 0.0,
    scale = 1.0,
    rotation = 0.0,
  }, Texture)
end

function Texture.load(path, callback)
  if type(callback) == "function" then
    return ResourceAsync.request("texture", path, nil, function()
      return load_sync(path)
    end, callback)
  end

  return load_sync(path)
end

function Texture:draw(tint)
  if self.handle == nil or self.handle == 0 then
    return
  end

  draw_texture(self.handle,
               self.x, self.y,
               self.scale,
               self.rotation,
               tint or COLOR_WHITE)
end

function Texture:destroy()
  if self.handle ~= nil and self.handle ~= 0 then
    destroy_texture(self.handle)
    self.handle = 0
  end
end

return Texture

File: examples/remote/server/src/texture.ts
import type { ResourceManager } from "./resource_manager";
import {
  rl_texture_create,
  rl_texture_destroy,
  rl_texture_draw_ex,
  rl_texture_draw_ground,
  set_resource_manager as set_texture_resource_manager,
} from "./rl/rl_texture";

export class Texture {
  handle: number | null = null;

  constructor(
    public readonly path: string,
    private readonly resourceManager: ResourceManager,
  ) {}

  static async load(resourceManager: ResourceManager, path: string): Promise<Texture> {
    const texture = new Texture(path, resourceManager);
    set_texture_resource_manager(resourceManager);
    texture.handle = await rl_texture_create(path);
    return texture;
  }

  drawEx(x: number, y: number, scale: number, rotation: number, tint: number): void {
    if (this.handle == null) {
      return;
    }

    rl_texture_draw_ex(this.handle, x, y, scale, rotation, tint);
  }

  drawGround(positionX: number, positionY: number, positionZ: number, width: number, length: number, tint: number): void {
    if (this.handle == null) {
      return;
    }

    rl_texture_draw_ground(this.handle, positionX, positionY, positionZ, width, length, tint);
  }

  async destroy(): Promise<void> {
    if (this.handle == null) {
      return;
    }
    set_texture_resource_manager(this.resourceManager);
    await rl_texture_destroy(this.handle);
    this.handle = 0;
  }
}
